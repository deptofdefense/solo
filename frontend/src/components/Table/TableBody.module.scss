
.td {
  background-color: #f0f0f0;
  border-color: white;
}

.subComponentTd {
  background-color: white;
  border-color: white;
}

