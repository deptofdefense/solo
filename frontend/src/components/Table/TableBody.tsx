import React, { PropsWithChildren } from "react";
import { TableInstance, Row } from "react-table";
import classes from "./TableBody.module.scss";

interface TableBodyProps<T extends object> extends TableInstance<T> {
  renderSubComponent?: (row: Row<T>) => JSX.Element;
}

const TableBody = <T extends object>({
  rows,
  visibleColumns,
  prepareRow,
  renderSubComponent,
  getTableBodyProps
}: PropsWithChildren<TableBodyProps<T>>) => {
  return (
    <tbody {...getTableBodyProps()}>
      {rows.map(row => {
        prepareRow(row);
        const { key, ...rest } = row.getRowProps();
        return (
          <React.Fragment key={key}>
            <tr {...rest}>
              {row.cells.map(cell => {
                const { key, ...rest } = cell.getCellProps();
                return (
                  <td key={key} {...rest} className={classes.td}>
                    {cell.render("Cell")}
                  </td>
                );
              })}
            </tr>

            {row.isExpanded && renderSubComponent ? (
              <tr>
                <td
                  colSpan={visibleColumns.length}
                  className={classes.subComponentTd}
                >
                  {renderSubComponent(row)}
                </td>
              </tr>
            ) : null}
          </React.Fragment>
        );
      })}
    </tbody>
  );
};

export default TableBody;
