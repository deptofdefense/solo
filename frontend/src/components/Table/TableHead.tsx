import React, { PropsWithChildren } from "react";
import { HeaderGroup } from "react-table";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faChevronDown, faChevronUp } from "@fortawesome/free-solid-svg-icons";

interface TableHeadProps<T extends object> {
  headerGroups: HeaderGroup<T>[];
}

const TableHead = <T extends object>({
  headerGroups
}: PropsWithChildren<TableHeadProps<T>>) => (
  <thead>
    {headerGroups.map(headerGroup => {
      const { key, ...rest } = headerGroup.getHeaderGroupProps();
      return (
        <tr {...rest} key={key}>
          {headerGroup.headers.map(column => {
            const { key, ...rest } = column.getHeaderProps(
              column.getSortByToggleProps()
            );
            return (
              <th {...rest} key={key}>
                <div className="display-flex flex-justify flex-align-center flex-no-wrap text-no-wrap">
                  {column.render("Header")}
                  {!column.disableSortBy && (
                    <FontAwesomeIcon
                      icon={column.isSortedDesc ? faChevronDown : faChevronUp}
                      style={{
                        visibility: column.isSorted ? "visible" : "hidden"
                      }}
                      className="margin-left-05"
                    />
                  )}
                </div>
              </th>
            );
          })}
        </tr>
      );
    })}
  </thead>
);

export default TableHead;
