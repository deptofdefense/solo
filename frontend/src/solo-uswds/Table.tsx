import React from "react";
import classNames from "classnames";

interface USWDSTableProps extends Partial<JSX.IntrinsicElements["table"]> {
  borderless?: boolean;
}

export const Table: React.FC<USWDSTableProps> = ({
  children,
  borderless = true,
  ...rest
}) => (
  <table
    className={classNames("usa-table", "grid-col-12", {
      "usa-table--borderless": borderless
    })}
    {...rest}
  >
    {children}
  </table>
);
