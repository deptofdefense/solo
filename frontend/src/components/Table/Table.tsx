import React, { useEffect } from "react";
import {
  useTable,
  useExpanded,
  Row,
  Column,
  useSortBy,
  SortingRule
} from "react-table";
import { Table as USWDSTable, TableData } from "solo-uswds";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faChevronUp, faChevronDown } from "@fortawesome/free-solid-svg-icons";

interface TableProps<T extends object> {
  columns: Column<T>[];
  data: T[];
  initialSortBy: SortingRule<T>;
  onSort: (sortBy: SortingRule<T>) => void;
  renderSubComponent?: (row: Row<T>) => JSX.Element;
}

function Table<T extends object>({
  columns,
  data,
  initialSortBy,
  onSort,
  renderSubComponent
}: TableProps<T>) {
  const {
    getTableProps,
    getTableBodyProps,
    headerGroups,
    rows,
    prepareRow,
    visibleColumns,
    state: { sortBy }
  } = useTable(
    {
      columns,
      data,
      manualSortBy: true,
      initialState: { sortBy: [initialSortBy] }
    },
    useSortBy,
    useExpanded
  );

  useEffect(() => {
    onSort(sortBy[0]);
  }, [sortBy, onSort]);

  return (
    <>
      <USWDSTable {...getTableProps()}>
        <thead>
          {headerGroups.map(headerGroup => (
            <tr {...headerGroup.getHeaderGroupProps()}>
              {headerGroup.headers.map(column => (
                <th {...column.getHeaderProps(column.getSortByToggleProps())}>
                  <div className="display-flex flex-justify flex-align-center flex-no-wrap text-no-wrap">
                    {column.render("Header")}
                    <FontAwesomeIcon
                      icon={column.isSortedDesc ? faChevronDown : faChevronUp}
                      style={{
                        visibility: column.isSorted ? "visible" : "hidden"
                      }}
                      className="margin-left-05"
                    />
                  </div>
                </th>
              ))}
            </tr>
          ))}
        </thead>
        <tbody {...getTableBodyProps()}>
          {rows.map(row => {
            prepareRow(row);
            const { key, ...rest } = row.getRowProps();
            return (
              <React.Fragment key={key}>
                <tr {...rest}>
                  {row.cells.map(cell => (
                    <TableData {...cell.getCellProps()}>
                      {cell.render("Cell")}
                    </TableData>
                  ))}
                </tr>

                {row.isExpanded && renderSubComponent ? (
                  <tr>
                    <TableData colSpan={visibleColumns.length} details>
                      {renderSubComponent(row)}
                    </TableData>
                  </tr>
                ) : null}
              </React.Fragment>
            );
          })}
        </tbody>
      </USWDSTable>
    </>
  );
}

export default Table;
