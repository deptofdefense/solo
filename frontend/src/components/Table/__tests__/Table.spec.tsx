import React from "react";
import { Column } from "react-table";
import Table from "../Table";
import { render, fireEvent, wait } from "test-utils";

interface TestDoc {
  value: any;
}

describe("Table component", () => {
  const columns: Column<TestDoc>[] = [
    {
      Header: "expand",
      Cell: ({ row }) => (
        <span data-testid="testexpand" {...row.getToggleRowExpandedProps()}>
          expand
        </span>
      )
    },
    {
      Header: "test column",
      accessor: "value"
    }
  ];

  const data = [{ value: "2" }];
  const onSortMock = jest.fn();
  const onRenderSubMock = jest.fn();

  afterEach(() => {
    onSortMock.mockReset();
    onRenderSubMock.mockReset();
  });

  it("matches snapshot", () => {
    const { asFragment } = render(
      <Table<TestDoc>
        columns={columns}
        data={data}
        initialSortBy={{ id: "value" }}
        onSort={onSortMock}
        renderSubComponent={onRenderSubMock}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("calls onSort callback on sort toggled", async () => {
    const { getByText } = render(
      <Table<TestDoc>
        columns={columns}
        data={data}
        initialSortBy={{ id: "value" }}
        onSort={onSortMock}
        renderSubComponent={onRenderSubMock}
      />
    );
    await wait(() => {
      expect(onSortMock).toHaveBeenCalled();
    });
    const sortHeader = getByText("test column");
    fireEvent.click(sortHeader);
    await wait(() => {
      expect(onSortMock).toHaveBeenCalledTimes(2);
      expect(onSortMock.mock.calls[1][0]).toMatchObject({
        id: "value"
      });
    });
    fireEvent.click(sortHeader);
    await wait(() => {
      expect(onSortMock).toHaveBeenCalledTimes(3);
      expect(onSortMock.mock.calls[2][0]).toBeUndefined();
    });
  });

  it("calls render subcomponent on click", async () => {
    const { getByTestId } = render(
      <Table<TestDoc>
        columns={columns}
        data={data}
        initialSortBy={{ id: "value" }}
        onSort={onSortMock}
        renderSubComponent={onRenderSubMock}
      />
    );
    const toggleBtn = getByTestId("testexpand");
    fireEvent.click(toggleBtn);
    await wait(() => {
      expect(onRenderSubMock).toHaveBeenCalled();
    });
  });
});
