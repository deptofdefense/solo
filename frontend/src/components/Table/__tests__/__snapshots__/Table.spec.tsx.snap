// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table component matches snapshot 1`] = `
<DocumentFragment>
  <table
    class="usa-table grid-col-12 usa-table--borderless"
    role="table"
  >
    <thead>
      <tr
        role="row"
      >
        <th
          colspan="1"
          role="columnheader"
        >
          <div
            class="display-flex flex-justify flex-align-center flex-no-wrap text-no-wrap"
          >
            expand
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-chevron-up fa-w-14 margin-left-05"
              data-icon="chevron-up"
              data-prefix="fas"
              focusable="false"
              role="img"
              style="visibility: hidden;"
              viewBox="0 0 448 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M240.971 130.524l194.343 194.343c9.373 9.373 9.373 24.569 0 33.941l-22.667 22.667c-9.357 9.357-24.522 9.375-33.901.04L224 227.495 69.255 381.516c-9.379 9.335-24.544 9.317-33.901-.04l-22.667-22.667c-9.373-9.373-9.373-24.569 0-33.941L207.03 130.525c9.372-9.373 24.568-9.373 33.941-.001z"
                fill="currentColor"
              />
            </svg>
          </div>
        </th>
        <th
          colspan="1"
          role="columnheader"
          style="cursor: pointer;"
          title="Toggle SortBy"
        >
          <div
            class="display-flex flex-justify flex-align-center flex-no-wrap text-no-wrap"
          >
            test column
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-chevron-up fa-w-14 margin-left-05"
              data-icon="chevron-up"
              data-prefix="fas"
              focusable="false"
              role="img"
              style="visibility: visible;"
              viewBox="0 0 448 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M240.971 130.524l194.343 194.343c9.373 9.373 9.373 24.569 0 33.941l-22.667 22.667c-9.357 9.357-24.522 9.375-33.901.04L224 227.495 69.255 381.516c-9.379 9.335-24.544 9.317-33.901-.04l-22.667-22.667c-9.373-9.373-9.373-24.569 0-33.941L207.03 130.525c9.372-9.373 24.568-9.373 33.941-.001z"
                fill="currentColor"
              />
            </svg>
          </div>
        </th>
      </tr>
    </thead>
    <tbody
      role="rowgroup"
    >
      <tr
        role="row"
      >
        <td
          role="cell"
          style="background-color: rgb(240, 240, 240); border-color: white;"
        >
          <span
            data-testid="testexpand"
            style="cursor: pointer;"
            title="Toggle Row Expanded"
          >
            expand
          </span>
        </td>
        <td
          role="cell"
          style="background-color: rgb(240, 240, 240); border-color: white;"
        >
          2
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
